x-pack/plugins/data_enhanced/server/search/session/session_service.ts

```diff
@@ -0,0 +1,363 @@
+
+      updatedSessions.forEach((updatedSavedObject) => {
+        const sessionInfo = this.sessionSearchMap.get(updatedSavedObject.id)!;
+        if (updatedSavedObject.error) {
+          // Retry next time
+          sessionInfo.retryCount++;
+        } else if (updatedSavedObject.attributes.idMapping) {
+          // Delete the ids that we just saved, avoiding a potential new ids being lost.
+          Object.keys(updatedSavedObject.attributes.idMapping).forEach((key) => {
+            sessionInfo.ids.delete(key);
+          });
+          // If the session object is empty, delete it as well
+          if (!sessionInfo.ids.entries.length) {
+            this.sessionSearchMap.delete(updatedSavedObject.id);
+          } else {
+            sessionInfo.retryCount = 0;
+          }
+        }
+      });
+    } catch (e) {
+      this.logger.error(`monitorMappedIds | Error fetching sessions. ${e}`);
+    }
+  }
+
+  private async updateAllSavedObjects(
+    activeMappingObjects: Array<SavedObject<BackgroundSessionSavedObjectAttributes>>
+  ) {
+    if (!activeMappingObjects.length) return [];
+
+    this.logger.debug(`updateAllSavedObjects | Updating ${activeMappingObjects.length} items`);
+    const updatedSessions = activeMappingObjects
+      ?.filter((so) => !so.error)
+      .map((sessionSavedObject) => {
+        const sessionInfo = this.sessionSearchMap.get(sessionSavedObject.id);
+        const idMapping = sessionInfo ? Object.fromEntries(sessionInfo.ids.entries()) : {};
+        sessionSavedObject.attributes.idMapping = {
+          ...sessionSavedObject.attributes.idMapping,
+          ...idMapping,
+        };
+        return sessionSavedObject;
+      });
+
+    const updateResults = await this.internalSavedObjectsClient.bulkUpdate<BackgroundSessionSavedObjectAttributes>(
+      updatedSessions
+    );
+    return updateResults.saved_objects;
+  }
+
+  public search<Request extends IKibanaSearchRequest, Response extends IKibanaSearchResponse>(
+    strategy: ISearchStrategy<Request, Response>,
+    searchRequest: Request,
+    options: ISearchOptions,
+    searchDeps: SearchStrategyDependencies,
+    deps: BackgroundSessionDependencies
+  ): Observable<Response> {
+    // If this is a restored background search session, look up the ID using the provided sessionId
+    const getSearchRequest = async () =>
+      !options.isRestore || searchRequest.id
+        ? searchRequest
+        : {
+            ...searchRequest,
+            id: await this.getId(searchRequest, options, deps),
+          };
+
+    return from(getSearchRequest()).pipe(
+      switchMap((request) => strategy.search(request, options, searchDeps)),
+      tapFirst((response) => {
+        if (searchRequest.id || !options.sessionId || !response.id || options.isRestore) return;
+        this.trackId(searchRequest, response.id, options, deps);
+      })
+    );
+  }
+
+  // TODO: Generate the `userId` from the realm type/realm name/username
```

Comment:
Check with us before doing so. We are working with the ES security team to come up with a unique user identifier that doesn't rely on the realm name.